src/translatador.cpp

```diff
@@ -0,0 +1,385 @@
+        sentences.reserve(expectedCount * 2);
+        sentence_ids.reserve(expectedCount * 2);
+    }
+
+    void push(bergamot::Segments&& segments, bergamot::AnnotatedText&& source) {
+        sources.push_back(std::move(source));
+        sentence_counts.push_back(segments.size());
+
+        for (bergamot::Segment& sentence : segments) {
+            if (sentence.size() > max_length) {
+                max_length = sentence.size();
+            }
+            sentences.push_back(std::move(sentence));
+            sentence_ids.push_back(sentence_id++);
+        }
+    }
+};
+
+struct ModelData {
+    std::shared_ptr<marian::Options> options;
+    bergamot::AlignedMemory model_memory;
+    // short_list_generator holds a raw reference to this memory
+    bergamot::AlignedMemory short_list_memory;
+    bergamot::Vocabs vocabs;
+    bergamot::TextProcessor text_processor;
+    std::shared_ptr<marian::data::BinaryShortlistGenerator> short_list_generator;
+
+    [[nodiscard]] bergamot::Vocabs create_vocabs(const Buffer source_vocab, const Buffer target_vocab) const {
+        static constexpr size_t ALIGNMENT = 64;
+        if (source_vocab != target_vocab && target_vocab) {
+            return bergamot::Vocabs(options, {
+                std::make_shared<bergamot::AlignedMemory>(source_vocab.aligned_copy(ALIGNMENT)),
+                std::make_shared<bergamot::AlignedMemory>(target_vocab.aligned_copy(ALIGNMENT))
+            });
+        } else {
+            std::shared_ptr<bergamot::AlignedMemory> vocab_memory = std::make_shared<bergamot::AlignedMemory>(source_vocab.aligned_copy(ALIGNMENT));
+            return bergamot::Vocabs(options, {vocab_memory, vocab_memory});
+        }
+    }
+
+    std::shared_ptr<marian::data::BinaryShortlistGenerator> create_short_list_generator() {
+        if (short_list_memory.begin() && short_list_memory.size() > 0) {
+            bool shared = vocabs.sources().front() == vocabs.target();
+            return std::make_shared<marian::data::BinaryShortlistGenerator>(
+                short_list_memory.begin(), short_list_memory.size(),
+                vocabs.sources().front(), vocabs.target(),
+                0, 1, shared, false
+            );
+        }
+        return {};
+    }
+
+    ModelData(
+        std::shared_ptr<marian::Options> options,
+        const Buffer model,
+        const Buffer source_vocab,
+        const Buffer target_vocab,
+        const Buffer short_list
+    ): options(std::move(options)),
+       model_memory(model.aligned_copy(256)),
+       short_list_memory(short_list.aligned_copy(64)),
+       vocabs(create_vocabs(source_vocab, target_vocab)),
+       text_processor(bergamot::TextProcessor(this->options, this->vocabs, bergamot::AlignedMemory())),
+       short_list_generator(create_short_list_generator()) {
+    }
+
+    ModelData(const ModelData&) = delete;
+
+    ModelData& operator=(const ModelData&) = delete;
+
+    ~ModelData() {
```

Comment:
Since the destructor is defined and copy constructor/assignment are deleted, you should implement the move constructor/assignment. See C++'s rule of 5